diffedit.md

```diff
@@ -0,0 +1,752 @@
+            pred = unet(input, t, encoder_hidden_states=text_embeddings).sample
+
+        # perform guidance
+        pred_uncond, pred_text = pred.chunk(2)
+        pred_nonscaled = pred_uncond + guidance_scale * (pred_text - pred_uncond)
+        pred = pred_nonscaled
+
+        # compute the "previous" noisy sample
+        if (next_step < num_inference_steps):
+            xt = create_noise(original_image_scheduler, original_image_latents, start_step=next_step)
+        else:
+            xt = original_image_latents
+
+        yt = denoising_scheduler.step(pred, t, noisy_latents).prev_sample
+
+        noisy_latents = mask*yt+((1-mask)*xt)
+
+    return noisy_latents
+
+
+def diffedit(original_image, caption, query, *args, **kwargs):
+    mask = create_mask(original_image, caption, query, *args, **kwargs)
+    plot_mask(original_image, mask)
+    final_latents = masked_image_generation(original_image, query, mask, *args, **kwargs)
+    images_to_grid([original_image, to_image(final_latents)], 2)
+```
+
+Now our function is defined we can test it out! One of the prompts that the DiffEdit paper first shows is changing a bowl of fruits to a bowl of pears, so we'll also demonstrate that here:
+
+
+```python
+img_path = '/kaggle/input/fruitbowl/fruit_holding.png'
+fruit_bowl_image = Image.open(img_path).resize((512, 512))
+fruit_bowl_image
+```
+
+
+
+
+    
+![png](diffedit_files/diffedit_91_0.png)
+    
+
+
+
+
+```python
+caption = "a bowl of fruits"
+query = "a bowl of pears"
+
+diffedit(fruit_bowl_image, caption, query, noise_strength=1)
+```
+
+
+      0%|          | 0/50 [00:00<?, ?it/s]
+
+
+
+    
+![png](diffedit_files/diffedit_92_1.png)
+    
+
+
+
+    
+![png](diffedit_files/diffedit_92_2.png)
+    
+
+
+We can see a very nice mask generation here, along with some slightly wonky looking pears, that's mainly an issue with our diffusion model's image generation in this case rather than anything that we've done specifically, so we'll accept slightly wonky pears for now.
+
+## Limitations
+
+Something else which is worth considering is the limitations of the DiffEdit mask creation technique. If there are huge differences between the original image caption and the query string then the mask that is generated is not as accurate, which has the knock-on effect of leading to poor final image generation. Take our example below, we'll use the same original image and caption as before, but this time use the string `"bowl of cars"` as the query. As we can see the mask creation includes random areas of the background due to the lack of similarities with the original image and the intended query, and generates a rather odd final image.
+
+
+```python
+caption = "bowl of fruits"
+query = "bowl of cars"
+
+diffedit(fruit_bowl_image, caption, query, noise_strength=1)
+```
+
+
+      0%|          | 0/50 [00:00<?, ?it/s]
+
+
+
+    
+![png](diffedit_files/diffedit_95_1.png)
+    
+
+
+
+    
+![png](diffedit_files/diffedit_95_2.png)
+    
+
```

Comment:
You need a conclusion to ease the reader's landing back into the real world. Random ideas for a conclusion, pick and choose as you see fit -
We set out to do this, it worked as/not as we expected. It was more/less interesting/challenging/fun that expected. If I did it again I would do/not do.... In doing this work I came across this other area which I'm interested in investigating.

Reply:
I've now added a conclusion. Let me know if that's enough or if there's anything else you think we should add

Reply:
LGTM